src/ol/render/webgl/BatchRenderer.js

```diff
@@ -0,0 +1,204 @@
+   * the invert transform of the webgl buffers
+   * @param {import("./MixedGeometryBatch.js").GeometryBatch} batch Geometry batch
+   * @param {import("../../transform.js").Transform} currentTransform Transform
+   * @param {import("../../PluggableMap.js").FrameState} frameState Frame state.
+   * @param {number} offsetX X offset
+   */
+  render(batch, currentTransform, frameState, offsetX) {
+    // multiply the current projection transform with the invert of the one used to fill buffers
+    this.helper_.makeProjectionTransform(frameState, currentTransform);
+    translateTransform(currentTransform, offsetX, 0);
+    multiplyTransform(currentTransform, batch.invertVerticesBufferTransform);
+
+    // enable program, buffers and attributes
+    this.helper_.useProgram(this.program_, frameState);
+    this.helper_.bindBuffer(batch.verticesBuffer);
+    this.helper_.bindBuffer(batch.indicesBuffer);
+    this.helper_.enableAttributes(this.attributes_);
+
+    const renderCount = batch.indicesBuffer.getSize();
+    this.helper_.drawElements(0, renderCount);
+  }
+
+  /**
+   * Rebuild rendering instructions based on the provided frame state
+   * This is specific to the geometry type and has to be implemented by subclasses.
+   * @param {import("./MixedGeometryBatch.js").GeometryBatch} batch Geometry batch
+   * @protected
+   */
+  generateRenderInstructions_(batch) {
+    abstract();
+  }
+
+  /**
+   * Rebuild internal webgl buffers for rendering based on the current rendering instructions;
+   * This is asynchronous: webgl buffers wil _not_ be updated right away
+   * @param {import("./MixedGeometryBatch.js").GeometryBatch} batch Geometry batch
+   * @param {import("../../geom/GeometryType.js").default} geometryType Geometry type
+   * @param {function(): void} callback Function called once the render buffers are updated
+   * @protected
+   */
+  generateBuffers_(batch, geometryType, callback) {
+    const messageId = workerMessageCounter++;
+
+    let messageType;
+    switch (geometryType) {
+      case GeometryType.POLYGON:
+        messageType = WebGLWorkerMessageType.GENERATE_POLYGON_BUFFERS;
+        break;
+      case GeometryType.POINT:
+        messageType = WebGLWorkerMessageType.GENERATE_POINT_BUFFERS;
+        break;
+      case GeometryType.LINE_STRING:
+        messageType = WebGLWorkerMessageType.GENERATE_LINE_STRING_BUFFERS;
+        break;
+      default:
+      // pass
+    }
+
+    /** @type {import('./constants.js').WebGLWorkerGenerateBuffersMessage} */
+    const message = {
+      id: messageId,
+      type: messageType,
+      renderInstructions: batch.renderInstructions.buffer,
+      renderInstructionsTransform: batch.renderInstructionsTransform,
+      customAttributesCount: this.customAttributes_.length,
+    };
+    // additional properties will be sent back as-is by the worker
+    message['projectionTransform'] = batch.renderInstructionsTransform;
```

Comment:
Curious about the bracket notation here.  Are the other message properties getting renamed?

Reply:
Good catch, this was a useless leftover from the points layer renderer (originally I used a bracket notation for type checking reasons, i.e. setting a property that was not part of the defined type).